src/implicit.rs

```diff
@@ -487,27 +487,17 @@ fn identify_bracket_pairs<D: BidiDataSource>(
     let mut stack = vec![];
 
     for (run_index, level_run) in run_sequence.runs.iter().enumerate() {
-        let slice = if let Some(slice) = text.get(level_run.clone()) {
-            slice
-        } else {
-            #[cfg(feature = "std")]
-            std::debug_assert!(
-                false,
-                "Found broken indices in level run: found indices {}..{} for string of length {}",
-                level_run.start,
-                level_run.end,
-                text.len()
-            );
-            return ret;
-        };
-
-        for (i, ch) in slice.char_indices() {
-            let actual_index = level_run.start + i;
+        let mut actual_index = level_run.start;
+
+        while actual_index < level_run.end {
+            let (ch, len) = text.char_at(actual_index).unwrap();
```

Comment:
issue: no panics, please.

The older structure of the code was intentional to use `char_indices`, please retain something similar. I think we can add an indexing method that returns a `&Self` to `TextSource` and then call `char_indices()` on it